### ISSUE TITLE
minor changes to hoon.hoon lead to bail

### ISSUE THREAD

**Issue by joshuareagan** [urbit/urbit#1071]
tl:dr -- I keep making minor changes to hoon.hoon and find myself able to cause `bail`s too easily.  What follows is one example of this:

In hoon.hoon I changed some lead cores `^?` to iron `^|` (the ones with samples, and hence which should be iron) and did nothing else.  On a fresh release-candidate.  I then tried to `|reset` two times and got the following `bail-fail` on the second attempt:

```
[%vega-start-hoon /~zod/home/~2018.11.17..22.52.08..071e/sys/hoon]
%vega-parsed
%vega-compiled
[%vega-arvo /~zod/home/~2018.11.17..22.52.08..071e/sys/arvo]
%vega-minted
%arvo-assembly
%arvo-assembled
[%vega-kernel 0x71b5.6de2]
[%vega-compiled 141 141]
%hoon-load
[%tang /~zod/home/~2018.11.17..22.52.08..071e/sys/zuse ~sardef-dornyl]
[%vane %a /~zod/home/~2018.11.17..22.52.08..071e/sys/vane/ames ~dospen-bilmud]
[%vane %b /~zod/home/~2018.11.17..22.52.08..071e/sys/vane/behn ~locmex-fonber]
[%vane %c /~zod/home/~2018.11.17..22.52.08..071e/sys/vane/clay ~torben-sonsug]
[%vane %d /~zod/home/~2018.11.17..22.52.08..071e/sys/vane/dill ~ladrem-nodfyl]
[%vane %e /~zod/home/~2018.11.17..22.52.08..071e/sys/vane/eyre ~tidbep-ronlyr]
[%vane %f /~zod/home/~2018.11.17..22.52.08..071e/sys/vane/ford ~lardex-bidwes]
[%vane %g /~zod/home/~2018.11.17..22.52.08..071e/sys/vane/gall ~nammus-pildun]
[%vane %j /~zod/home/~2018.11.17..22.52.08..071e/sys/vane/jael ~mastes-fabsev]
~zod:dojo> |reset|
bail: foul
bailing out
urbit: ../noun/manage.c:653: u3m_bail: Assertion `0' failed.
                                                            Aborted (core dumped)
```

I have seen this happen a lot lately when I've been working on hoon.hoon.  Something feels wrong and brittle.  Either I'm repeatedly doing something stupid (definitely possible) or something is broken.

**Comment by joemfb** [comment]
Transferred to urbit/urbit as this appears to be a reference-counting bug. 

@joshuareagan, I have a workaround: for now, disable autoload and reset manually.

@frodwith, does anything jump out to you about this backtrace?

```
  * frame #0: 0x00007fff6b037b86 libsystem_kernel.dylib`__pthread_kill + 10
    frame #1: 0x00007fff6b0edc50 libsystem_pthread.dylib`pthread_kill + 285
    frame #2: 0x00007fff6afa11c9 libsystem_c.dylib`abort + 127
    frame #3: 0x00007fff6af69868 libsystem_c.dylib`__assert_rtn + 320
    frame #4: 0x0000000104795592 urbit`u3m_bail(how=1819635558) at manage.c:653 [opt]
    frame #5: 0x000000010478a9d6 urbit`u3a_take [inlined] _me_gain_use(dog=<unavailable>) at allocate.c:987 [opt]
    frame #6: 0x000000010478a9cc urbit`u3a_take [inlined] _me_take_north(dog=<unavailable>) at allocate.c:1250 [opt]
    frame #7: 0x000000010478a9cc urbit`u3a_take(som=<unavailable>) at allocate.c:1295 [opt]
    frame #8: 0x0000000104791c7f urbit`u3j_site_copy(dst_u=0x0000000207c0adc6, src_u=0x000000027f76875c, los_o='\0') at jets.c:1395 [opt]
    frame #9: 0x000000010479bd96 urbit`_n_prog_take_dat(dst_u=0x0000000207c0a610, src_u=0x000000027f768170, los_o='\0') at nock.c:2397 [opt]
    frame #10: 0x0000000104798a16 urbit`_n_reap(kev=<unavailable>) at nock.c:2473 [opt]
    frame #11: 0x000000010478e77c urbit`_ch_walk_node(han_u=0x000000027d610038, lef_w=20, fun_f=(urbit`_ch_walk_plain at hashtable.c:840), wit=0x00000001047989a0) at hashtable.c:795 [opt]
    frame #12: 0x000000010478e804 urbit`u3h_walk at hashtable.c:831 [opt]
    frame #13: 0x000000010478e7bb urbit`u3h_walk(har_p=<unavailable>, fun_f=(urbit`_n_reap at nock.c:2459)) at hashtable.c:850 [opt]
    frame #14: 0x0000000104795b19 urbit`u3m_love(pro=3245079066) at manage.c:834 [opt]
    frame #15: 0x0000000104795f18 urbit`u3m_soft_top(sec_w=<unavailable>, pad_w=1048576, fun_f=(urbit`u3v_poke at vortex.c:388), arg=3241656702) at manage.c:944 [opt]
    frame #16: 0x00000001047965af urbit`u3m_soft(sec_w=<unavailable>, fun_f=<unavailable>, arg=<unavailable>) at manage.c:1180 [opt]
    frame #17: 0x0000000104779516 urbit`u3_raft_work [inlined] _raft_punk(ovo=<unavailable>) at raft.c:1576 [opt]
    frame #18: 0x0000000104779478 urbit`u3_raft_work at raft.c:1993 [opt]
    frame #19: 0x0000000104778173 urbit`u3_lo_shut(inn='\0') at loop.c:446 [opt]
    frame #20: 0x000000010478211f urbit`_term_read_cb [inlined] _term_suck(uty_u=<unavailable>, buf=<unavailable>, siz_i=<unavailable>) at term.c:731 [opt]
    frame #21: 0x0000000104781cbe urbit`_term_read_cb(tcp_u=0x00007f8ff051d320, siz_i=<unavailable>, buf_u=0x00007ffeeb4894f8) at term.c:742 [opt]
    frame #22: 0x0000000104bcbbb4 libuv.1.dylib`uv__stream_io at stream.c:1192 [opt]
    frame #23: 0x0000000104bcb7b8 libuv.1.dylib`uv__stream_io(loop=<unavailable>, w=<unavailable>, events=<unavailable>) at stream.c:1259 [opt]
    frame #24: 0x0000000104bd4374 libuv.1.dylib`uv__io_poll(loop=0x0000000104bd8840, timeout=27000) at kqueue.c:276 [opt]
    frame #25: 0x0000000104bc3403 libuv.1.dylib`uv_run(loop=0x0000000104bd8840, mode=UV_RUN_DEFAULT) at core.c:354 [opt]
    frame #26: 0x00000001047890b1 urbit`main(argc=8, argv=0x00007ffeeb4927f8) at main.c:697 [opt]
    frame #27: 0x00007fff6aef9085 libdyld.dylib`start + 1
    frame #28: 0x00007fff6aef9085 libdyld.dylib`start + 1
```

**Comment by frodwith** [comment]
Well, there's a reference counting error somewhere. We're trying to take a call site's label and it has a zero reference count, I think? The label is only ever seen or touched by the jet system, so the error is probably near a call into that system (u3j_gate_slam, u3j_hank_..., etc) or within the jet system itself. Can't really say more without a gdb adventure.

**Comment by cgyarvin** [comment]
I think it’s time for someone to have a gdb adventure with Josh...

**Comment by joshuareagan** [comment]
I am willing to go on such an adventure.

**Comment by joshuareagan** [comment]
@joemfb I turned off `|autoload` and am getting the same bail as before.  I'll keep playing around with it though...

**Comment by joemfb** [comment]
Can you be more specific? Is this on a fresh fake-zod? How many resets before the bail occurs? What hoon changes? etc.

**Comment by joshuareagan** [comment]
yes, fresh fakezod, did a mount, turned off autoload, changed hoon.hoon only by replacing the same few `^?` with `^|`, reset once went fine, second reset led to bail.  I'm playing around with order to see how much that matters.

**Comment by joshuareagan** [comment]
Same result when I turned off autoload before mounting.  This is on Ubuntu 16, I'll try again on the mac to see whether that matters

**Comment by joshuareagan** [comment]
Mac behavior is indeed different -- two `|reset`s don't bail, but on the third I do get the bail.

**Comment by ohAitch** [comment]
This may be a stretch, but can you get this up as a travis test? See
https://github.com/urbit/arvo/blob/master/.travis/get-or-build-pill.sh for
a starting point on interleaving UNIX commands and urbit dojo input
If the pier restarts jinx it, maybe just boot with a hoon2.hoon in sys/,
and |cp it from dojo. (Or named pipe shenanigans into a backgrounded
urbit-runner?)

**Comment by joemfb** [comment]
I can picture this as a temporary test, but I'm not sure how to add it to the harness and make it work longterm (without having to keep two slightly-different copies of hoon.hoon in sync). If it's truly just the reset, I guess we could inject a comment at the beginning of the file.

**Comment by frodwith** [comment]
I'm growing more confident that this is a problem with the way memory is managed in call sites. Call sites for non-jetted, located cores seemingly mishandle memory - at least by losing an extra label. I don't have a fix yet, but I'm narrowing in.

**Comment by ohAitch** [comment]
Yeah, I definitely meant it just to pin down the reproducibility while
debugging this issue, not as a long-term test. (Tho checking in a .patch
with the ^? change would probably work, mostly I'm concerned about net test
runtime)

**Comment by frodwith** [comment]
Pretty sure I've nailed it down. Nothing terribly interesting, just an order-of-operations problem (should reap call sites before reaping the warm dashboard). Pull request incoming after I make a clean patch and test it.

**Comment by joshuareagan** [comment]
Nice work, much appreciated!

**Comment by frodwith** [comment]
https://github.com/urbit/urbit/pull/1072

**Comment by joshuareagan** [comment]
Problem solved with PR by @frodwith.
